src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -236,13 +236,13 @@ export class CallbackWorkflow implements Workflow {
       // exists, then fallback to the request callback.
       if (refreshCallback) {
         context.refreshToken = entry.tokenResult.refreshToken;
-        result = await refreshCallback(serverInfo, context);
+        result = await withLock(refreshCallback)(serverInfo, context);
       } else {
-        result = await requestCallback(serverInfo, context);
+        result = await withLock(requestCallback)(serverInfo, context);
       }
     } else {
       // With no token in the cache we use the request callback.
-      result = await requestCallback(serverInfo, context);
+      result = await withLock(requestCallback)(serverInfo, context);
```

Comment:
This calls the callbacks with a new lock every time – you probably do need to store the lock as part of the cache entry (or that’s how I’m reading the spec text)